docs/src/userguide/freethreading.rst

```diff
@@ -160,3 +160,298 @@ to require extreme care to make it work correctly).
+
+* avoid the Python-call overhead of the threading module,
+* use atomic variables to update numeric types in a controlled way without locking.
+
+For this our recommendation is to use the C++ standard library.  Most of these
+are available simply by "cimporting" from ``libcpp``.  In the event that Cython
+hasn't already wrapped what you want to use then you can do it yourself - our
+``libcpp`` is provided for convenience but it does nothing that can't be done
+with regular Cython code.
+
+The C standard library also provides some of these features (e.g. atomic variables
+and mutexes).  However, compiler support for the C++ standard library is better
+(in particular for MSVC) and the C++ standard library is more fully featured,
+so we recommend this first.
+
+One difficulty is with types that are not default constructable or moveable
+(e.g. ``latch``, ``semaphore``, ``barrier``).  These are difficult to
+stack-allocate because of how Cython's code-genertion works, so you
+need to heap-allocate them::
+
+  from libcpp.latch cimport latch
+
+  l = new latch(2)
+  try:
+    ...  # use the latch
+  finally:
+    del l
+
+It is also possible to use C++ to create new threads (for example, using the ``std::jthread``
+class).  This works, but we generally recommend creating threads through Python
+instead, mainly because the outermost ``with gil:`` statement in a non-Python
+created thread will not work reliably with multiple subinterpreters - this recommendation is
+therefore mainly to future-proofing your code and not restrict where it can
+be used from.  It is a fairly soft suggestion though, so feel free to ignore
+it if you have good reason to.
+
+``cython.critical_section`` vs GIL
+----------------------------------
+
+Understanding what protection a ``critical_section`` provides is
+important to being able to use it safely,  and it's also worth comparing
+it to the guarantees that the GIL provides.  Unfortunately some of
+this is very much an implementation detail of Python at the moment, so
+may be subject to change.
+
+What is guaranteed to be safe for both of ``critical_section`` and
+the GIL (on non-freethreading builds) is reading and writing to
+``cdef`` attributes of extension types::
+
+  cdef class C:
+    cdef object attr
+
+  ...
+  
+  cdef C c_instance = C()
+  with cython.critical_section(c_instance):
+    c_instance.attr = something
+
+  with cython.critical_section(c_instance):
+    something = c_instance.attr
+
+In principle, both a ``critical_section`` and the GIL can be interrupted
+by executing arbitrary Python code.  Arbitrary Python code can notable
+include the finalizers of any objects being destroyed.  This means that
+reassigning a Python attribute can trigger arbitrary code (but typically
+only after the new value has been put in place).  Additionally, triggering
+the GC can result in arbitrary code being executed (which means any Python
+memory allocation can trigger it).
```

Comment:
TODO: Memory allocation can trigger the GC in Python <3.11. Still worth noting but should caveat it.